x-pack/plugins/cross_cluster_replication/server/plugin.ts

```diff
@@ -115,12 +130,10 @@ export class CrossClusterReplicationServerPlugin implements Plugin<void, void, a
       }
     );
 
-    // Extend the elasticsearchJs client with additional endpoints.
-    const esClientConfig = { plugins: [elasticsearchJsPlugin] };
-    const ccrEsClient = elasticsearch.createClient('crossClusterReplication', esClientConfig);
-    http.registerRouteHandlerContext('crossClusterReplication', (ctx, request) => {
+    http.registerRouteHandlerContext('crossClusterReplication', async (ctx, request) => {
+      const client = await this.getCustomEsClient(getStartServices);
```

Comment:
This comment applies to the other files changed in this PR, too. Can we remove the side-effect of setting `this.ccrEsClient` within `getCustomEsClient`? I think it would be easier to follow this code if it's a pure function. I'd also like to add a comment so people understand why this function is extracted. Something like this (note that this function can also be extracted from the plugin class):

```ts
// TODO: Remove this once we are using the new elasticsearch-js client
const getCustomEsClient = async (getStartServices: CoreSetup['getStartServices']) => {
  const [core] = await getStartServices();
  // Extend the elasticsearchJs client with additional endpoints.
  const esClientConfig = { plugins: [elasticsearchJsPlugin] };
  return core.elasticsearch.legacy.createClient(
    'crossClusterReplication',
    esClientConfig
  );
}

// snip

    http.registerRouteHandlerContext('crossClusterReplication', async (ctx, request) => {
      this.ccrEsClient = await this.getCustomEsClient(getStartServices);
      return {
        client: this.ccrEsClient.asScoped(request),
      };
    });
```

Reply:
@cjcenizal this will create a new client on *every* request. It's quite an expensive operation because the platform needs to parse the config. https://github.com/elastic/kibana/blob/56041f03ada9730e701508dbc10032b69de9f218/src/core/server/elasticsearch/cluster_client.ts#L148

I can move client creation to the `start`:
```js
setup(core){
   http.registerRouteHandlerContext('crossClusterReplication', (ctx, request) => {
      return {
        client: this.ccrEsClient!.asScoped(request),
      };
    });
}
start(core){
  const esClientConfig = { plugins: [elasticsearchJsPlugin] };
  this.ccrEsClient = core.elasticsearch.legacy.createClient(
    'crossClusterReplication',
    esClientConfig
  );
}
```
Do you prefer this approach?

Reply:
Oh, I didn't realize it was called on each request! Thanks for explaining that. I think your suggestion is an improvement on the original, but it does seem counter-intuitive for a handler defined in `setup` to have a dependency defined by `start`. Let me toss one more option to you... 😄 

How about in my original suggestion we use the cached `ccrEsClient` if it exists, otherwise create a new one?

```ts
    http.registerRouteHandlerContext('crossClusterReplication', async (ctx, request) => {
      this.ccrEsClient = this.ccrEsClient ?? await this.getCustomEsClient(getStartServices);
      return {
        client: this.ccrEsClient.asScoped(request),
      };
    });
```

Reply:
ok, will do tomorrow